Fix for Cross Site Scripting Rule vulnerabilities

sumi-mathew opened the pull request:
## Description

 This PR  addressing a Cross-Site Scripting (XSS) vulnerability discovered during a static scan.

## Motivation and Context
Fixing XSS vulnerabilities is critical to securing web applications and protecting both user data and website integrity.

## Impact
<!---Describe any public API or user-facing feature change or any performance impact-->

## Test Plan
<!---Please fill in how you tested your change-->

## Contributor checklist

- [ ] Please make sure your submission complies with our [development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), [commit message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests), and [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution).
- [ ] PR description addresses the issue accurately and concisely.  If the change is non-trivial, a GitHub Issue is referenced.
- [ ] Documented new properties (with its default value), SQL syntax, functions, or other functionality.
- [ ] If release notes are required, they follow the [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).
- [ ] Adequate tests were added if applicable.
- [ ] CI passed.

## Release Notes
Please follow [release notes guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) and fill in the release notes below.

```
== RELEASE NOTES ==

Security
* Fix Cross Site Scripting Rule vulnerabilities. :pr:`24029`

yingsu00 commented on presto-main/src/main/java/com/facebook/presto/server/AsyncPageTransportServlet.java:
```diff
@@ -181,6 +183,7 @@ public void onError(AsyncEvent event)
             {
                 String errorMessage = format("Server error to process task result request %s : %s", requestURI, event.getThrowable().getMessage());
                 log.error(event.getThrowable(), errorMessage);
+                errorMessage = StringEscapeUtils.escapeHtml4(errorMessage);
```
Inline 186 to 187

sumi-mathew replied:
The comments have been addressed.

yingsu00 reviewed (commented):
Release note:

`* Fix provided for Cross Site Scripting Rule vulnerabilities.   :pr:`24029`
--->
`* Fix Cross Site Scripting Rule vulnerabilities.   :pr:`24029`

steveburnett commented:
Minor nits for the release note entry. 

```
== RELEASE NOTES ==

Security Changes
* Fix Cross Site Scripting Rule vulnerabilities. :pr:`24029`

```

tdcmeehan commented on presto-main/pom.xml:
```diff
@@ -36,6 +36,11 @@
             <artifactId>commons-math3</artifactId>
         </dependency>
 
+        <dependency>
```
Is there any way to do this without adding a dependency on commons?

sumi-mathew replied:
The security scanner recommends using a proper library for encoding output (such as the OWASP ESAPI Encoding module, OWASP Java HTML Sanitizer, OWASP AntiSamy, or Apache Shiro). In the IBM Pesto main module, the org.apache.commons.lang3 dependency is already included, so there is no need to add any additional dependencies for this fix. I also tried to apply the same changes for OSS, as this dependency is already used by other modules.

If I address the question, the answer is yes, we can manually validate special characters. While manual escaping works, it can become difficult when handling more complex HTML/JS edge cases. Libraries like StringEscapeUtils are more robust and thoroughly tested.

tdcmeehan commented on presto-main/src/main/java/com/facebook/presto/server/AsyncPageTransportServlet.java:
```diff
@@ -116,7 +116,44 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
+        StringBuilder stringBuilder = new StringBuilder(message.length());
+
+        // Traverse each character in the string and replace only when necessary
+        for (int i = 0; i < message.length(); i++) {
```
Please put this in a utility class and add testing for this utility class.

sumi-mathew replied:
The comments have been addressed.

steveburnett commented:
Thanks for the release note entry! A few formatting notes to follow the [Release Note Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines). 

```
== RELEASE NOTES ==

Security Changes
* Fix Cross Site Scripting Rule vulnerabilities. :pr:`24029`
```

tdcmeehan reviewed (commented):
LGTM % nit

sumi-mathew commented on presto-main/src/main/java/com/facebook/presto/server/AsyncPageTransportServlet.java:
```diff
@@ -116,9 +117,8 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response)
             throws IOException
     {
-        response.sendError(SC_BAD_REQUEST, message);
+        response.sendError(SC_BAD_REQUEST, SanitizeMessage.getSanitizeMessage(message));
```
Addressed the changes

tdcmeehan commented:
Can you let me know how the XSS attack would work here?  Give an example.

sumi-mathew commented:
> Can you let me know how the XSS attack would work here? Give an example.

From  contract  scan details:
 Software places user-controlled input in page content. An attacker could inject browser script code that is executed in the client browser. The end-user is the attacked subject, and the software is the vehicle for the attack. There are two main kinds of XSS: * Reflected XSS: Attacker causes victim to supply malicious content to a vulnerable web application, which renders HTML content embedding a malicious script executed in victim's browser. A variation of this is named DOM-based XSS, where the vulnerable software does not generate content depending on user input but include script code that use user-controlled input. * Persisted XSS: Attacker provides malicious content to vulnerable application. When other user access to vulnerable pages that embed without proper neutralization the attacker content, script code is executed in the victim's browser. The script executed in the victim's browser could perform malicious activities. Many browsers could limit the damage via security restrictions (e.g. 'same origin policy'), but user browsers generally allow scripting languages (e.g. JavaScript) in their browsers (disabling JavaScript severely limits a web site).

As per the suggestion from the contrast scan, I used a library to validate the message and executed the security scan again. The issue was resolved after that fix. I then raised the OSS PR, and as per your suggestion, I removed the library and added manual validation.

sumi-mathew commented:
As discussed ,It seems that the only attack vector is the incoming URL.  already validate the URL in the parseURI method.
So closing this PR.
